transformer_engine/jax/attention.py

```diff
@@ -215,11 +214,6 @@ def make_helper(attn_mask_type):
     if not make_helper(attn_mask_type).is_fused_attn_kernel_available():
         return False
 
-    # For context parallel need to check additional masking types
```

Comment:
So there is no way to pass `None` from maxtext when CP is not used for the axis? I think we should keep this check here so that we can avoid exceptions on cuDNN when we don't have appropriate support.

Reply:
Hi @mgoldfarb-nvidia , we can definitely pass a boolean from jax side. However, I decided to remove this because:
1. Wanted to make as minimal changes as possible to the high level `DotProductAttention` api
2. The same check is performed [here](https://github.com/NVIDIA/TransformerEngine/blob/7fb22c375804f77f4f95df3eab606c7bd3e80aed/transformer_engine/jax/cpp_extensions/attention.py#L1044). It essentially provides meaningful error msg and its higher than cudnn level:
```
...
...
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/workspace/maxtext_workspace/context/maxtext/MaxText/train.py", line 757, in <module>
    app.run(main)
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 308, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 254, in _run_main
    sys.exit(main(argv))
  File "/opt/workspace/maxtext_workspace/context/maxtext/MaxText/train.py", line 753, in main
    train_loop(config)
  File "/opt/workspace/maxtext_workspace/context/maxtext/MaxText/train.py", line 649, in train_loop
    state, metrics = p_train_step(state, example_batch, nextrng)
jaxlib.xla_extension.XlaRuntimeError: INTERNAL: custom_partitioner: Traceback (most recent call last):
  File "/opt/workspace/maxtext_workspace/context/maxtext/MaxText/train.py", line 757, in <module>
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 308, in run
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 254, in _run_main
  File "/opt/workspace/maxtext_workspace/context/maxtext/MaxText/train.py", line 753, in main
  File "/opt/workspace/maxtext_workspace/context/maxtext/MaxText/train.py", line 649, in train_loop
  File "/opt/jax/jax/_src/traceback_util.py", line 180, in reraise_with_filtered_traceback
  File "/opt/jax/jax/_src/pjit.py", line 337, in cache_miss
  File "/opt/jax/jax/_src/pjit.py", line 187, in _python_pjit_helper
  File "/opt/jax/jax/_src/core.py", line 2820, in bind
  File "/opt/jax/jax/_src/core.py", line 442, in bind_with_trace
  File "/opt/jax/jax/_src/core.py", line 955, in process_primitive
  File "/opt/jax/jax/_src/pjit.py", line 1736, in _pjit_call_impl
  File "/opt/jax/jax/_src/pjit.py", line 1712, in call_impl_cache_miss
  File "/opt/jax/jax/_src/pjit.py", line 1642, in _pjit_call_impl_python
  File "/opt/jax/jax/_src/interpreters/pxla.py", line 2346, in compile
  File "/opt/jax/jax/_src/interpreters/pxla.py", line 2855, in from_hlo
  File "/opt/jax/jax/_src/interpreters/pxla.py", line 2667, in _cached_compilation
  File "/opt/jax/jax/_src/compiler.py", line 434, in compile_or_get_cached
  File "/opt/jax/jax/_src/compiler.py", line 662, in _compile_and_write_cache
  File "/opt/jax/jax/_src/profiler.py", line 333, in wrapper
  File "/opt/jax/jax/_src/compiler.py", line 267, in backend_compile
  File "/opt/jax/jax/_src/custom_partitioning.py", line 155, in _custom_partitioning_partition
  File "/opt/transformer-engine/transformer_engine/jax/cpp_extensions/attention.py", line 1202, in partition
  File "/opt/transformer-engine/transformer_engine/jax/cpp_extensions/attention.py", line 1046, in check_supported
ValueError: Context parallel fused attention only supports masking types:  NVTE_Mask_Type.NVTE_NO_MASK,NVTE_Mask_Type.NVTE_CAUSAL_MASK got: NVTE_Mask_Type.NVTE_PADDING_CAUSAL_MASK
```

Reply:
Okay- I have a proposal although it might not be the cleanest. We use this method in our unit tests to also know if the configuration is supported before running i.e. to skip invalid configs we know to be failing.

We could add the `is_context_parallel` argument back but as an `Optional[bool]` and if its passed as `None` we don't attempt to do the more specific check. That way we can still query the full support at the top level

Reply:
Can we infer `is_context_parallel` from `cp_axis`? I noticed there is a similar logic [here](https://github.com/NVIDIA/TransformerEngine/blob/c036765b1eb611c1b7e8db72676f93b36e9ee36d/transformer_engine/jax/cpp_extensions/attention.py#L1194). If so, we can pass `is_context_parallel` [here](https://github.com/NVIDIA/TransformerEngine/blob/v1.11/transformer_engine/jax/flax/transformer.py#L523) in the `DotProductAttention` module without adding an additional argument.

Reply:
The challenge @kocchop ran into is that this check only works once the code is being transformed by Jax jit. He found that in MaxText the axis information is not available at the DPA api level and thus the implicit axis check fails.

Reply:
Got it. Then how about his
We keep `is_fused_attn_kernel_available` context_parallism-agnostic, means no `is_context_parallel` argument. But we call `is_fused_attn_kernel_available` again in `_FusedAttnCPWithAllGatherHelper.check_support` if the `attn_mask_type == AttnMaskType.CAUSAL_MASK`.
Unlike other config, we can still fall back to the unfused attn, if we don't have the fused attn kernels with CP, then we can just raise a ValueError for that

Reply:
That works and maybe the best compromise here. The one place we need to update then is the unit test which needs to skip configs not supported. We currently rely on `is_fused_attn_kernel_available` to do this check.

We can call into `_FusedAttnCPWithAllGatherHelper.check_support` as you suggest here from the unit test to also properly skip as needed.